### PR TITLE
1431/google translate

### PR DESCRIPTION
**Pull request by software-project**
# Pull Request Template

Addresses # (#1431 )

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

Adds Google translations to GET listing API endpoint. 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Prototype/POC (not to merge)
- [ ] This change is a refactor/address technical debt
- [ ] This change requires a documentation update
- [ ] This change requires a SQL Script

## How Can This Be Tested/Reviewed?

Please describe the tests that you ran to verify your changes. Provide instructions so we can review. Please also list any relevant details for your test configuration

- [x] Desktop View
- [x] Mobile View
- [ ] Test A
- [ ] Test B

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [x] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [x] I have updated the changelog to include a description of my changes

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -29,6 +29,7 @@ All notable changes to this project will be documented in this file. The format
   - convert listing address jsonb columns to separate address tables
   - removed unused inverse relations from entities
   - recreated foreign keys constraints for `application_flagged_set_applications_applications`, `listings_leasing_agents_user_accounts`, `property_group_properties_property` and add missing `NOT NULL` migration for listing name column
+  - add google translations for GET listing API call
```
Comment (emilyjablonski) [inline CHANGELOG.md:32]: Can you add the right formatting here? It should look like `- description (([#1590](https://github.com/bloom-housing/bloom/pull/1590)) (contributor name)`

**Comment by emilyjablonski** [comment]
Tried to test this locally and got a `Request failed with status code 500` when trying to switch to Spanish. I re-ran yarn install and re-seeded. Is there some other setup I need to do? Did you maybe need to push changes to the env_template files for `GOOGLE_API_KEY` ?

File: backend/core/src/listings/listings.controller.ts [hunk 2]
```diff
@@ -102,12 +103,15 @@ export class ListingsController {
 
   @Get(`:listingId`)
   @ApiOperation({ summary: "Get listing by id", operationId: "retrieve" })
-  @UseInterceptors(CacheInterceptor)
-  async retrieve(@Param("listingId") listingId: string): Promise<ListingDto> {
+  // @UseInterceptors(CacheInterceptor)
```
Comment (jaredcwhite) [inline backend/core/src/listings/listings.controller.ts:106]: Was this left commented out to debug something? Or did you mean to change the behavior?

File: backend/core/src/listings/listings.service.ts [hunk 3]
```diff
@@ -137,6 +162,66 @@ export class ListingsService {
...
+        client_email: process.env.GOOGLE_API_EMAIL,
+      },
+      projectId: process.env.GOOGLE_API_ID,
+    })
+  }
+  // Sets value to the object by string path. eg. "property.accessibility" or "preferences.0.title"
+  private setValue = (object, path, value) =>
+    path
+      .split(".")
+      .reduce((o, p, i) => (o[p] = path.split(".").length === ++i ? value : o[p] || {}), object)
```
Comment (jaredcwhite) [inline backend/core/src/listings/listings.service.ts:195]: Can we use more verbose variable names here? Hard to parse what it's doing… 😅

File: backend/core/src/listings/listings.service.ts [hunk 4]
```diff
@@ -137,6 +162,66 @@ export class ListingsService {
...
+      // Attach translated values to the listing
+      translations.keys.forEach((key, index) => {
+        this.setValue(result, key, translatedValues[0][index])
+      })
+    }
+
+    return result
+  }
+
+  private translateService = () => {
```
Comment (jaredcwhite) [inline backend/core/src/listings/listings.service.ts:182]: I like the approach you've laid out here, but I'm wondering if everything from this point to the end of the file should be abstracted out of the listings service into its own module or whatever. That way we could generically use Google Translate from anywhere we might want in the future (not just listings).

**Review by jaredcwhite — CHANGES_REQUESTED** [review]
@software-project Cool stuff! Had a question and a couple suggestions.

File: CHANGELOG.md (hunk 1 above)
Comment (software-project) [reply CHANGELOG.md]: Sure and done.

File: backend/core/src/listings/listings.controller.ts (hunk 2 above)
Comment (software-project) [reply backend/core/src/listings/listings.controller.ts:107]: That's on purpose. I've added a TODO and a follow up ticket: https://github.com/bloom-housing/bloom/issues/1636

File: backend/core/src/listings/listings.service.ts (hunk 3 above)
Comment (software-project) [reply backend/core/src/listings/listings.service.ts]: `o` and `p` corresponded with `object` and `path` but I've renamed them to be more direct as requested.

File: backend/core/src/listings/listings.service.ts (hunk 4 above)
Comment (seanmalbert) [reply backend/core/src/listings/listings.service.ts:182]: @software-project ,
I agree with @jaredcwhite and you set this up in a way that should allow you to easily do that. Can you create a `/src/translation` dir that has a `translation.service.ts` file with the appropriate code transferred to it. getTranslations can continue to receive listing: Listing, but in the future, if we ever need it for anything else, we can update it to use a generic.

**Comment by seanmalbert** [comment]
@software-project and @emilyjablonski , I added a custom cache interceptor, `ListingLangCacheInterceptor` to handle caching the language routes. Can you please review that piece?

**Comment by seanmalbert** [comment]
We may find that a different cache strategy would work better than caching the entire result for each language, especially since there's overlap. So we could update the interceptor to get the listing from the main cache, and cache just the translations separately and add them back on the data.

**Comment by emilyjablonski** [comment]
cache changes lgtm from code review perspective 👍

**Comment by software-project** [comment]
@seanmalbert I can't approve, as I'm the author of this PR, but caching changes are looking good and build is working as expected 👍

**Review by emilyjablonski — APPROVED** [review]
